Mod2/Lessons/Week2/TDD.md

```diff
@@ -0,0 +1,152 @@
+
+> With a partner, take 5 minutes to determine the smallest amount of coding we need to do in order to resolve this first issue.
+
+<details><summary>Solution</summary><br/>
+The only thing we need to do is create an Item class (and add the project reference).  Doing that will clear the first issue, and presents us with new errors!
+
+![](/Mod2/Images/Week2/ItemClassCreated.png)
+</details>
+
+One of the biggest benefits of TDD is that we _only create the code that we need_.  Following TDD should ensure that we are not creating unnecessary code.  But, to get this benefit, we need to make **small, incremental** changes.  You should repeat the following steps:
+
+1. Identify a single issue
+2. Make the smallest change possible to fix that issue
+3. Repeat step 1
+
+>With your partner, follow this incremental process to complete the Item class.  Keep track of how many errors/issues you come across and be ready to share out!
+
+** Instructor Note ** Be sure to ask them at what point they started running the test to see what to do next.  It is possible (likely) that students have not run the tests at all - they may rely only on the text editor's feedback to determine what to do next. Following the steps above, they _should_ get to a point where the properties are created but are not assigned to the constructor argument values.
+
+<details><summary>Completed Item Class</summary><br/>
+![](/Mod2/Images/Week2/ItemClassComplete.png)
+</details>
+
+
+### TDD the Customer and Order Class
+
+Using TDD often shows us how interdependent our classes can be. This forces us to imagine not only what the classes will look like, but also how they will work together.  
+
+> With a partner, create the tests for the Customer class and Order class.
+> As you work, reach out with **any** questions!
+
+** Instructor Note ** As you review groups' code, keep an eye on issues and bring the group back together to discuss common misunderstandings. When students are ready to move on to the Order, help them brainstorm what the Item List might look like.
+
+<details><summary>Completed Customer and Order Tests</summary><br/> 
+Order Test
+![](/Mod2/Images/Week2/OrderTest.png)
+
+Customer Test
+![](/Mod2/Images/Week2/CustomerTest.png)
+</details>
+
+> With a partner, use your tests to drive the implementation of the Customer and Order classes.
+
+
+## Types of Tests
+
+When writing a program, you will likely have smaller methods that support each other to create greater functionality.  Often, these might be wrapped in some kind of runner method, or chained together for a grand result.  The tests for these methods are different kinds of tests.  There are four commonly referred to types of tests that build upon each other:
```

Comment:
I think probably best to talk about this in a later mod, but at some point I think it would be great to talk about more broadly types of testing. Such as load testing, penetration testing/security related testing, manual testing as apposed to automated, usability/accessibility testing, disaster recovery testing, help expand our student's definitions of what's out there with regards to software development.